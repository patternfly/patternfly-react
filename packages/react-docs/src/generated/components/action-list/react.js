import React from 'react';
import { AutoLinkHeader, Example, Link as PatternflyThemeLink } from 'theme-patternfly-org/components';
import TimesIcon from '@patternfly/react-icons/dist/js/icons/times-icon';
import CheckIcon from '@patternfly/react-icons/dist/js/icons/check-icon';
const pageData = {
  "id": "Action list",
  "section": "components",
  "source": "react",
  "slug": "/components/action-list/react",
  "sourceLink": "https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/ActionList/examples/ActionList.md",
  "propComponents": [
    {
      "name": "ActionList",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list",
          "defaultValue": "''"
        },
        {
          "name": "isIconList",
          "type": "boolean",
          "description": "Flag indicating the action list contains multiple icons and item padding should be removed"
        }
      ]
    },
    {
      "name": "ActionListGroup",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list group"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list group",
          "defaultValue": "''"
        }
      ]
    },
    {
      "name": "ActionListItem",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list item"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list item",
          "defaultValue": "''"
        }
      ]
    }
  ],
  "cssPrefix": [
    "pf-c-action-list"
  ],
  "examples": [
    "Action list single group",
    "Action list with icons",
    "Action list multiple groups",
    "Action list with cancel button"
  ]
};
pageData.liveContext = {
  TimesIcon,
  CheckIcon
};
pageData.examples = {
  'Action list single group': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport {\n  ActionList,\n  ActionListItem,\n  Button,\n  Dropdown,\n  DropdownItem,\n  DropdownSeparator,\n  KebabToggle\n} from '@patternfly/react-core';\n\nexport const ActionListSingleGroup: React.FunctionComponent = () => {\n  const [isOpen, setIsOpen] = React.useState(false);\n\n  const onToggle = (\n    isOpen: boolean,\n    event: MouseEvent | TouchEvent | KeyboardEvent | React.KeyboardEvent<any> | React.MouseEvent<HTMLButtonElement>\n  ) => {\n    event.stopPropagation();\n    setIsOpen(isOpen);\n  };\n\n  const onSelect = (event: React.SyntheticEvent<HTMLDivElement, Event>) => {\n    event.stopPropagation();\n    setIsOpen(!isOpen);\n  };\n\n  const dropdownItems = [\n    <DropdownItem key=\"link\">Link</DropdownItem>,\n    <DropdownItem key=\"action\" component=\"button\">\n      Action\n    </DropdownItem>,\n    <DropdownItem key=\"disabled link\" isDisabled>\n      Disabled Link\n    </DropdownItem>,\n    <DropdownItem key=\"disabled action\" isDisabled component=\"button\">\n      Disabled Action\n    </DropdownItem>,\n    <DropdownSeparator key=\"separator\" />,\n    <DropdownItem key=\"separated link\">Separated Link</DropdownItem>,\n    <DropdownItem key=\"separated action\" component=\"button\">\n      Separated Action\n    </DropdownItem>\n  ];\n\n  return (\n    <React.Fragment>\n      <ActionList>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button\">\n            Back\n          </Button>\n        </ActionListItem>\n      </ActionList>\n      <br />\n      With kebab\n      <ActionList>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button2\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button2\">\n            Back\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Dropdown\n            onSelect={onSelect}\n            toggle={<KebabToggle onToggle={onToggle} />}\n            isOpen={isOpen}\n            isPlain\n            dropdownItems={dropdownItems}\n            position=\"right\"\n          />\n        </ActionListItem>\n      </ActionList>\n    </React.Fragment>\n  );\n};\n","title":"Action list single group","lang":"ts"}}>
      
    </Example>,
  'Action list with icons': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListItem, Button } from '@patternfly/react-core';\nimport TimesIcon from '@patternfly/react-icons/dist/esm/icons/times-icon';\nimport CheckIcon from '@patternfly/react-icons/dist/esm/icons/check-icon';\n\nexport const ActionListWithIcons: React.FunctionComponent = () => (\n  <ActionList isIconList>\n    <ActionListItem>\n      <Button variant=\"plain\" id=\"times-button\" aria-label=\"times icon button\">\n        <TimesIcon />\n      </Button>\n    </ActionListItem>\n    <ActionListItem>\n      <Button variant=\"plain\" id=\"check-button\" aria-label=\"check icon button\">\n        <CheckIcon />\n      </Button>\n    </ActionListItem>\n  </ActionList>\n);\n","title":"Action list with icons","lang":"ts"}}>
      
    </Example>,
  'Action list multiple groups': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListGroup, ActionListItem, Button } from '@patternfly/react-core';\n\nexport const ActionListMultipleGroups: React.FunctionComponent = () => (\n  <ActionList>\n    <ActionListGroup>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"next-button\">\n          Next\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"secondary\" id=\"back-button\">\n          Back\n        </Button>\n      </ActionListItem>\n    </ActionListGroup>\n    <ActionListGroup>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"submit-button\">\n          Submit\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"link\" id=\"cancel-button\">\n          Cancel\n        </Button>\n      </ActionListItem>\n    </ActionListGroup>\n  </ActionList>\n);\n","title":"Action list multiple groups","lang":"ts"}}>
      
    </Example>,
  'Action list with cancel button': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListGroup, ActionListItem, Button } from '@patternfly/react-core';\n\nexport const ActionListWithCancelButton: React.FunctionComponent = () => (\n  <React.Fragment>\n    In modals, forms, data lists\n    <ActionList>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"save-button\">\n          Save\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"link\" id=\"cancel-button\">\n          Cancel\n        </Button>\n      </ActionListItem>\n    </ActionList>\n    <br />\n    In wizards\n    <ActionList>\n      <ActionListGroup>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button\">\n            Back\n          </Button>\n        </ActionListItem>\n      </ActionListGroup>\n      <ActionListGroup>\n        <ActionListItem>\n          <Button variant=\"link\" id=\"cancel-button2\">\n            Cancel\n          </Button>\n        </ActionListItem>\n      </ActionListGroup>\n    </ActionList>\n  </React.Fragment>\n);\n","title":"Action list with cancel button","lang":"ts"}}>
      
    </Example>
};

const Component = () => (
  <React.Fragment>
    <AutoLinkHeader {...{"id":"examples","size":"h2","className":"ws-title ws-h2"}}>
      {`Examples`}
    </AutoLinkHeader>
    {React.createElement(pageData.examples["Action list single group"])}
    {React.createElement(pageData.examples["Action list with icons"])}
    {React.createElement(pageData.examples["Action list multiple groups"])}
    {React.createElement(pageData.examples["Action list with cancel button"])}
  </React.Fragment>
);
Component.displayName = 'ComponentsActionListReactDocs';
Component.pageData = pageData;

export default Component;
