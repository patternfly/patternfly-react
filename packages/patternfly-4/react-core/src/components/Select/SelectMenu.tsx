import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Select/select';
import { default as formStyles } from '@patternfly/react-styles/css/components/Form/form';
import { css } from '@patternfly/react-styles';
import { SelectConsumer, SelectVariant } from './selectConstants';
import { Omit } from '../../helpers/typeUtils';

// tslint:disable-next-line
const FocusTrap: any = require('focus-trap-react');

export interface SelectMenuProps extends Omit<React.HTMLProps<HTMLElement>, 'checked' | 'selected' | 'ref'> {
  /** Content rendered inside the SelectMenu */
  children: React.ReactElement[];
  /** Additional classes added to the SelectMenu control */
  className?: string;
  /** Flag indicating the Select is expanded */
  isExpanded: boolean;
  /** Flag indicating the Select options are grouped */
  isGrouped?: boolean;
  /** Currently selected option (for single, typeahead variants) */
  selected?: string | string[];
  /** Currently checked options (for checkbox variant) */
  checked?: string[];
  /** Internal flag for specifiying how the menu was opened */
  openedOnEnter?: boolean;
  /** Internal callback for ref tracking */
  sendRef?: (ref: React.ReactNode, index: number) => void;
  /** Internal callback for keyboard navigation */
  keyHandler?: (index: number, position: string) => void;
}

export class SelectMenu extends React.Component<SelectMenuProps> {
  static defaultProps = {
    className: '',
    isExpanded: false,
    isGrouped: false,
    openedOnEnter: false,
    selected: '',
    sendRef: Function.prototype,
    keyHandler: Function.prototype
  };

  extendChildren() {
    const { children, isGrouped } = this.props;
    if (isGrouped) {
      let index = 0;
      return React.Children.map(children, (group: React.ReactElement) =>
        React.cloneElement(group, {
          titleId: group.props.label.replace(/\W/g, '-'),
          children: group.props.children.map((option: React.ReactElement) => this.cloneOption(option, index++))
        })
      );
    }
    return React.Children.map(this.props.children, (child: React.ReactElement, index: number) =>
      this.cloneOption(child, index)
    );
  }

  cloneOption(child: React.ReactElement, index: number) {
    const { selected, sendRef, keyHandler } = this.props;
    const isSelected =
      selected && selected.constructor === Array
        ? selected && selected.includes(child.props.value)
        : selected === child.props.value;
    return React.cloneElement(child, {
      id: `${child.props.value}-${index}`,
      isSelected,
      sendRef,
      keyHandler,
      index
    });
  }

  extendCheckboxChildren(props: any) {
    const { children, isGrouped, checked, sendRef, keyHandler } = this.props;
    const { 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy } = props;
    if (isGrouped) {
      let index = 0;
      return React.Children.map(children, (group: React.ReactElement) =>
        React.cloneElement(group, {
          titleId: group.props.label.replace(/\W/g, '-'),
          children: (
            <fieldset aria-labelledby={group.props.label.replace(/\W/g, '-')} className={css(formStyles.formFieldset)}>
              {group.props.children.map((option: React.ReactElement) =>
                React.cloneElement(option, {
                  isChecked: checked && checked.includes(option.props.value),
                  sendRef,
                  keyHandler,
                  index: index++
                })
              )}
            </fieldset>
          )
        })
      );
    }
    return (
      <fieldset
        {...props}
        aria-label={ariaLabel}
        aria-labelledby={(!ariaLabel && ariaLabelledBy) || null}
        className={css(formStyles.formFieldset)}
      >
        {React.Children.map(children, (child: React.ReactElement, index: number) =>
          React.cloneElement(child, {
            isChecked: checked && checked.includes(child.props.value),
            sendRef,
            keyHandler,
            index
          })
        )}
      </fieldset>
    );
  }

  render() {
    const {
      children,
      className,
      isExpanded,
      openedOnEnter,
      selected,
      checked,
      isGrouped,
      sendRef,
      keyHandler,
      ...props
    } = this.props;

    return (
      <SelectConsumer>
        {({ variant }) => (
          <React.Fragment>
            {variant !== SelectVariant.checkbox && (
              <ul {...props} className={css(styles.selectMenu, className)} role="listbox">
                {this.extendChildren()}
              </ul>
            )}
            {variant === SelectVariant.checkbox && React.Children.count(children) > 0 && (
              <FocusTrap focusTrapOptions={{ clickOutsideDeactivates: true }}>
                <div className={css(styles.selectMenu, className)}>
                  <form noValidate className={css(formStyles.form)}>
                    <div className={css(formStyles.formGroup)}>{this.extendCheckboxChildren(props)}</div>
                  </form>
                </div>
              </FocusTrap>
            )}
            {variant === SelectVariant.checkbox && React.Children.count(children) === 0 && (
              <div className={css(styles.selectMenu, className)}>
                <form noValidate className={css(formStyles.form)}>
                  <div className={css(formStyles.formGroup)}/>
                </form>
              </div>
            )}
          </React.Fragment>
        )}
      </SelectConsumer>
    );
  }
}
