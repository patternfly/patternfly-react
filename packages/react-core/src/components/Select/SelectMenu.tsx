import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Select/select';
import formStyles from '@patternfly/react-styles/css/components/Form/form';
import { css } from '@patternfly/react-styles';
import { SelectOptionObject, SelectOption } from './SelectOption';
import { SelectConsumer, SelectVariant } from './selectConstants';
import { PickOptional } from '../../helpers/typeUtils';

import { SelectGroup } from './SelectGroup';
import { Divider } from '../Divider/Divider';

export interface SelectMenuProps extends Omit<React.HTMLProps<HTMLElement>, 'checked' | 'selected' | 'ref'> {
  /** Content rendered inside the SelectMenu */
  children: React.ReactElement[] | React.ReactNode;
  /** Flag indicating that the children is custom content to render inside the SelectMenu.  If true, variant prop is ignored. */
  isCustomContent?: boolean;
  /** Additional classes added to the SelectMenu control */
  className?: string;
  /** Flag indicating the Select is expanded */
  isExpanded?: boolean;
  /** Flag indicating the Select options are grouped */
  isGrouped?: boolean;
  /** Currently selected option (for single, typeahead variants) */
  selected?: string | SelectOptionObject | (string | SelectOptionObject)[];
  /** Currently checked options (for checkbox variant) */
  checked?: (string | SelectOptionObject)[];
  /** Internal flag for specifiying how the menu was opened */
  openedOnEnter?: boolean;
  /** Flag to specify the  maximum height of the menu, as a string percentage or number of pixels */
  maxHeight?: string | number;
  /** Inner prop passed from parent */
  noResultsFoundText?: string;
  /** Inner prop passed from parent */
  createText?: string;
  /** Internal callback for ref tracking */
  sendRef?: (ref: React.ReactNode, favoriteRef: React.ReactNode, index: number) => void;
  /** Internal callback for keyboard navigation */
  keyHandler?: (index: number, innerIndex: number, position: string) => void;
  /** Flag indicating select has an inline text input for filtering */
  hasInlineFilter?: boolean;
  innerRef?: any;
}

class SelectMenuWithRef extends React.Component<SelectMenuProps> {
  static displayName = 'SelectMenu';
  static defaultProps: PickOptional<SelectMenuProps> = {
    className: '',
    isExpanded: false,
    isGrouped: false,
    openedOnEnter: false,
    selected: '',
    maxHeight: '',
    sendRef: () => {},
    keyHandler: () => {},
    isCustomContent: false,
    hasInlineFilter: false
  };

  extendChildren(randomId: string) {
    const { children, isGrouped } = this.props;
    const childrenArray: React.ReactElement[] = children as React.ReactElement[];
    if (isGrouped) {
      let index = 0;
      return React.Children.map(childrenArray, (group: React.ReactElement) => {
        if (group.type === SelectGroup) {
          return React.cloneElement(group, {
            titleId: group.props.label && group.props.label.replace(/\W/g, '-'),
            children: React.Children.map(group.props.children, (option: React.ReactElement) =>
              this.cloneOption(option, index++, randomId)
            )
          });
        } else {
          return this.cloneOption(group, index++, randomId);
        }
      });
    }
    return React.Children.map(childrenArray, (child: React.ReactElement, index: number) =>
      this.cloneOption(child, index, randomId)
    );
  }

  cloneOption(child: React.ReactElement, index: number, randomId: string) {
    const { selected, sendRef, keyHandler } = this.props;
    const isSelected = this.checkForValue(child.props.value, selected);
    if (child.type === Divider) {
      return child;
    }
    return React.cloneElement(child, {
      inputId: `${randomId}-${index}`,
      isSelected,
      sendRef,
      keyHandler,
      index
    });
  }

  checkForValue(
    valueToCheck: string | SelectOptionObject,
    options: string | SelectOptionObject | (string | SelectOptionObject)[]
  ) {
    if (!options || !valueToCheck) {
      return false;
    }

    const isSelectOptionObject =
      typeof valueToCheck !== 'string' &&
      (valueToCheck as SelectOptionObject).toString &&
      (valueToCheck as SelectOptionObject).compareTo;

    if (Array.isArray(options)) {
      if (isSelectOptionObject) {
        return options.some(option => (option as SelectOptionObject).compareTo(valueToCheck));
      } else {
        return options.includes(valueToCheck);
      }
    } else {
      if (isSelectOptionObject) {
        return (options as SelectOptionObject).compareTo(valueToCheck);
      } else {
        return options === valueToCheck;
      }
    }
  }

  extendCheckboxChildren(children: React.ReactElement[]) {
    const { isGrouped, checked, sendRef, keyHandler, hasInlineFilter } = this.props;
    let index = hasInlineFilter ? 1 : 0;
    if (isGrouped) {
      return React.Children.map(children, (group: React.ReactElement) => {
        if (group.type === SelectOption || group.type === Divider) {
          return group;
        }
        return React.cloneElement(group, {
          titleId: group.props.label && group.props.label.replace(/\W/g, '-'),
          children: (
            <fieldset
              aria-labelledby={group.props.label && group.props.label.replace(/\W/g, '-')}
              className={css(styles.selectMenuFieldset)}
            >
              {React.Children.map(group.props.children, (option: React.ReactElement) =>
                option.type === Divider
                  ? option
                  : React.cloneElement(option, {
                      isChecked: this.checkForValue(option.props.value, checked),
                      sendRef,
                      keyHandler,
                      index: index++
                    })
              )}
            </fieldset>
          )
        });
      });
    }
    return React.Children.map(children, (child: React.ReactElement) =>
      child.type === Divider
        ? child
        : React.cloneElement(child, {
            isChecked: this.checkForValue(child.props.value, checked),
            sendRef,
            keyHandler,
            index: index++
          })
    );
  }

  render() {
    /* eslint-disable @typescript-eslint/no-unused-vars */
    const {
      children,
      isCustomContent,
      className,
      isExpanded,
      openedOnEnter,
      selected,
      checked,
      isGrouped,
      sendRef,
      keyHandler,
      maxHeight,
      noResultsFoundText,
      createText,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      hasInlineFilter,
      innerRef,
      ...props
    } = this.props;
    /* eslint-enable @typescript-eslint/no-unused-vars */
    return (
      <SelectConsumer>
        {({ variant, inputIdPrefix }) => (
          <React.Fragment>
            {isCustomContent && (
              <div
                ref={innerRef}
                className={css(styles.selectMenu, className)}
                {...(maxHeight && { style: { maxHeight, overflow: 'auto' } })}
                {...props}
              >
                {children}
              </div>
            )}
            {variant !== SelectVariant.checkbox &&
              !isCustomContent &&
              (!isGrouped ? (
                <ul
                  ref={innerRef}
                  className={css(styles.selectMenu, className)}
                  role="listbox"
                  aria-label={ariaLabel}
                  aria-labelledby={(!ariaLabel && ariaLabelledBy) || null}
                  {...(maxHeight && { style: { maxHeight, overflow: 'auto' } })}
                  {...props}
                >
                  {this.extendChildren(inputIdPrefix)}
                </ul>
              ) : (
                <div
                  ref={innerRef}
                  className={css(styles.selectMenu, className)}
                  {...(maxHeight && { style: { maxHeight, overflow: 'auto' } })}
                  {...props}
                >
                  {this.extendChildren(inputIdPrefix)}
                </div>
              ))}
            {variant === SelectVariant.checkbox && !isCustomContent && React.Children.count(children) > 0 && (
              <div
                ref={innerRef}
                className={css(styles.selectMenu, className)}
                {...(maxHeight && { style: { maxHeight, overflow: 'auto' } })}
              >
                <fieldset
                  {...props}
                  aria-label={ariaLabel}
                  aria-labelledby={(!ariaLabel && ariaLabelledBy) || null}
                  className={css(formStyles.formFieldset)}
                >
                  {hasInlineFilter && [
                    (children as React.ReactElement[]).shift(),
                    ...this.extendCheckboxChildren(children as React.ReactElement[])
                  ]}
                  {!hasInlineFilter && this.extendCheckboxChildren(children as React.ReactElement[])}
                </fieldset>
              </div>
            )}
            {variant === SelectVariant.checkbox && !isCustomContent && React.Children.count(children) === 0 && (
              <div
                ref={innerRef}
                className={css(styles.selectMenu, className)}
                {...(maxHeight && { style: { maxHeight, overflow: 'auto' } })}
              >
                <fieldset className={css(styles.selectMenuFieldset)} />
              </div>
            )}
          </React.Fragment>
        )}
      </SelectConsumer>
    );
  }
}

export const SelectMenu = React.forwardRef((props, ref) => (
  <SelectMenuWithRef innerRef={ref} {...props}>
    {props.children}
  </SelectMenuWithRef>
));
