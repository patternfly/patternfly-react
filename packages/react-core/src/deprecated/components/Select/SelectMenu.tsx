import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Select/select';
import formStyles from '@patternfly/react-styles/css/components/Form/form';
import { css } from '@patternfly/react-styles';
import { SelectOptionObject, SelectOption } from './SelectOption';
import { SelectConsumer, SelectPosition, SelectVariant, SelectContextInterface } from './selectConstants';
import { PickOptional } from '../../../helpers/typeUtils';

import { SelectGroup } from './SelectGroup';
import { Divider } from '../../../components/Divider/Divider';

export interface SelectMenuProps extends Omit<React.HTMLProps<HTMLElement>, 'checked' | 'selected' | 'ref'> {
  /** Content rendered inside the SelectMenu */
  children: React.ReactElement[] | React.ReactNode;
  /** Flag indicating that the children is custom content to render inside the SelectMenu.  If true, variant prop is ignored. */
  isCustomContent?: boolean;
  /** Additional classes added to the SelectMenu control */
  className?: string;
  /** Flag indicating the Select is expanded */
  isExpanded?: boolean;
  /** Flag indicating the Select options are grouped */
  isGrouped?: boolean;
  /** Currently selected option (for single, typeahead variants) */
  selected?: string | SelectOptionObject | (string | SelectOptionObject)[];
  /** Currently checked options (for checkbox variant) */
  checked?: (string | SelectOptionObject)[];
  /** @hide Internal flag for specifiying how the menu was opened */
  openedOnEnter?: boolean;
  /** Flag to specify the  maximum height of the menu, as a string percentage or number of pixels */
  maxHeight?: string | number;
  /** Indicates where menu will be alligned horizontally */
  position?: SelectPosition | 'right' | 'left';
  /** Inner prop passed from parent */
  noResultsFoundText?: string;
  /** Inner prop passed from parent */
  createText?: string;
  /** @hide Internal callback for ref tracking */
  sendRef?: (ref: React.ReactNode, favoriteRef: React.ReactNode, index: number) => void;
  /** @hide Internal callback for keyboard navigation */
  keyHandler?: (index: number, innerIndex: number, position: string) => void;
  /** Flag indicating select has an inline text input for filtering */
  hasInlineFilter?: boolean;
  /** @hide forwarded ref */
  innerRef?: any;
  /** Content rendered in the footer of the select menu */
  footer?: React.ReactNode;
  /** The menu footer element */
  footerRef?: React.RefObject<HTMLDivElement>;
  /** @hide callback to check if option is the last one in the menu when there is a footer  */
  isLastOptionBeforeFooter?: (index: number) => void;
}

class SelectMenuWithRef extends React.Component<SelectMenuProps> {
  static displayName = 'SelectMenu';
  static defaultProps: PickOptional<SelectMenuProps> = {
    className: '',
    isExpanded: false,
    isGrouped: false,
    openedOnEnter: false,
    selected: '',
    maxHeight: '',
    position: SelectPosition.left,
    sendRef: () => {},
    keyHandler: () => {},
    isCustomContent: false,
    hasInlineFilter: false,
    isLastOptionBeforeFooter: () => {}
  };

  extendChildren(randomId: string) {
    const { children, hasInlineFilter, isGrouped } = this.props;
    const childrenArray: React.ReactElement[] = children as React.ReactElement[];
    let index = hasInlineFilter ? 1 : 0;
    if (isGrouped) {
      return React.Children.map(childrenArray, (group: React.ReactElement) => {
        if (group.type === SelectGroup) {
          return React.cloneElement(group, {
            titleId: group.props.label && group.props.label.replace(/\W/g, '-'),
            children: React.Children.map(group.props.children, (option: React.ReactElement) =>
              this.cloneOption(option, index++, randomId)
            )
          });
        } else {
          return this.cloneOption(group, index++, randomId);
        }
      });
    }
    return React.Children.map(childrenArray, (child: React.ReactElement) => this.cloneOption(child, index++, randomId));
  }

  cloneOption(child: React.ReactElement, index: number, randomId: string) {
    const { selected, sendRef, keyHandler } = this.props;
    const isSelected = this.checkForValue(child.props.value, selected);
    if (child.type === Divider) {
      return child;
    }
    return React.cloneElement(child, {
      inputId: `${randomId}-${index}`,
      isSelected,
      sendRef,
      keyHandler,
      index
    });
  }

  checkForValue(
    valueToCheck: string | SelectOptionObject,
    options: string | SelectOptionObject | (string | SelectOptionObject)[]
  ) {
    if (!options || !valueToCheck) {
      return false;
    }

    const isSelectOptionObject =
      typeof valueToCheck !== 'string' &&
      (valueToCheck as SelectOptionObject).toString &&
      (valueToCheck as SelectOptionObject).compareTo;

    if (Array.isArray(options)) {
      if (isSelectOptionObject) {
        return options.some((option) => (option as SelectOptionObject).compareTo(valueToCheck));
      } else {
        return options.includes(valueToCheck);
      }
    } else {
      if (isSelectOptionObject) {
        return (options as SelectOptionObject).compareTo(valueToCheck);
      } else {
        return options === valueToCheck;
      }
    }
  }

  extendCheckboxChildren(children: React.ReactElement[]) {
    const { isGrouped, checked, sendRef, keyHandler, hasInlineFilter, isLastOptionBeforeFooter } = this.props;
    let index = hasInlineFilter ? 1 : 0;
    if (isGrouped) {
      return React.Children.map(children, (group: React.ReactElement) => {
        if (group.type === Divider) {
          return group;
        } else if (group.type === SelectOption) {
          return React.cloneElement(group, {
            isChecked: this.checkForValue(group.props.value, checked),
            sendRef,
            keyHandler,
            index: index++,
            isLastOptionBeforeFooter
          });
        }
        return React.cloneElement(group, {
          titleId: group.props.label && group.props.label.replace(/\W/g, '-'),
          children: group.props.children ? (
            <fieldset
              aria-labelledby={group.props.label && group.props.label.replace(/\W/g, '-')}
              className={css(styles.selectMenuFieldset)}
            >
              {React.Children.map(group.props.children, (option: React.ReactElement) =>
                option.type === Divider
                  ? option
                  : React.cloneElement(option, {
                      isChecked: this.checkForValue(option.props.value, checked),
                      sendRef,
                      keyHandler,
                      index: index++,
                      isLastOptionBeforeFooter
                    })
              )}
            </fieldset>
          ) : null
        });
      });
    }
    return React.Children.map(children, (child: React.ReactElement) =>
      child.type === Divider
        ? child
        : React.cloneElement(child, {
            isChecked: this.checkForValue(child.props.value, checked),
            sendRef,
            keyHandler,
            index: index++,
            isLastOptionBeforeFooter
          })
    );
  }

  renderSelectMenu({ variant, inputIdPrefix }: SelectContextInterface) {
    /* eslint-disable @typescript-eslint/no-unused-vars */
    const {
      children,
      isCustomContent,
      className,
      isExpanded,
      openedOnEnter,
      selected,
      checked,
      isGrouped,
      position,
      sendRef,
      keyHandler,
      maxHeight,
      noResultsFoundText,
      createText,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      hasInlineFilter,
      innerRef,
      footer,
      footerRef,
      isLastOptionBeforeFooter,
      ...props
    } = this.props;
    /* eslint-enable @typescript-eslint/no-unused-vars */
    let Component = 'div';
    const variantProps = {
      ref: innerRef,
      className: css(
        !footer ? styles.selectMenu : 'pf-v5-c-select__menu-list',
        position === SelectPosition.right && styles.modifiers.alignRight,
        className
      ),
      ...(maxHeight && { style: { maxHeight, overflow: 'auto' } })
    } as React.HTMLAttributes<HTMLElement>;
    const extendedChildren = () =>
      variant === SelectVariant.checkbox
        ? this.extendCheckboxChildren(children as React.ReactElement[])
        : this.extendChildren(inputIdPrefix);

    if (isCustomContent) {
      variantProps.children = children;
    } else if (hasInlineFilter) {
      if (React.Children.count(children) === 0) {
        variantProps.children = <fieldset className={css(styles.selectMenuFieldset)} />;
      } else {
        variantProps.children = (
          <fieldset
            aria-label={ariaLabel}
            aria-labelledby={(!ariaLabel && ariaLabelledBy) || null}
            className={css(formStyles.formFieldset)}
          >
            {(children as React.ReactElement[]).shift()}
            {extendedChildren()}
          </fieldset>
        );
      }
    } else {
      variantProps.children = extendedChildren();
      if (!isGrouped) {
        Component = 'ul';
        variantProps.role = 'listbox';
        variantProps['aria-label'] = ariaLabel;
        variantProps['aria-labelledby'] = (!ariaLabel && ariaLabelledBy) || null;
      }
    }

    return (
      <React.Fragment>
        <Component {...variantProps} {...props} />
        {footer && (
          <div className={css(styles.selectMenuFooter)} ref={footerRef}>
            {footer}
          </div>
        )}
      </React.Fragment>
    );
  }

  render() {
    return <SelectConsumer>{(context) => this.renderSelectMenu(context)}</SelectConsumer>;
  }
}

export const SelectMenu = React.forwardRef<unknown, React.PropsWithChildren<unknown>>((props, ref) => (
  <SelectMenuWithRef innerRef={ref} {...props}>
    {props.children}
  </SelectMenuWithRef>
));
