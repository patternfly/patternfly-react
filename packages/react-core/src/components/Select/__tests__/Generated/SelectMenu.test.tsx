/**
 * This test was generated
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { SelectMenu } from '../../SelectMenu';
// any missing imports can usually be resolved by adding them here
import {} from '../..';

it('SelectMenu should match snapshot (auto-generated)', () => {
  const view = shallow(
    <SelectMenu
      children={[undefined as React.ReactElement]}
      isCustomContent={false}
      className={"''"}
      isExpanded={false}
      isGrouped={false}
      selected={''}
      checked={[]}
      openedOnEnter={false}
      maxHeight={''}
      noResultsFoundText={'string'}
      createText={'string'}
      sendRef={() => {}}
      keyHandler={() => {}}
    />
  );
  expect(view).toMatchSnapshot();
});
