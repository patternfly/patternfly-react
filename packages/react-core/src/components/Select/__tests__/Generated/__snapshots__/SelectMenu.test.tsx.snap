// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMenu should match snapshot (auto-generated) 1`] = `
<SelectMenu
  checked={Array []}
  className="''"
  createText="string"
  hasInlineFilter={false}
  innerRef={null}
  isCustomContent={false}
  isExpanded={false}
  isGrouped={false}
  isLastOptionBeforeFooter={[Function]}
  keyHandler={[Function]}
  maxHeight=""
  noResultsFoundText="string"
  openedOnEnter={false}
  selected=""
  sendRef={[Function]}
/>
`;
